src/lib/tls/tls_messages.h

```diff
@@ -548,23 +548,42 @@ class Certificate_Request_13;
 */
 class BOTAN_UNSTABLE_API Certificate_13 final : public Handshake_Message {
    public:
-      struct Certificate_Entry {
-            // TODO: RFC 8446 4.4.2 specifies the possibility to negotiate the usage
-            //       of a single raw public key in lieu of the X.509 certificate
-            //       chain. This is left for future work.
-            X509_Certificate certificate;
-            Extensions extensions;
+      class Certificate_Entry {
+         public:
+            Certificate_Entry(TLS_Data_Reader& reader, const Connection_Side side, const Certificate_Type cert_type);
+            Certificate_Entry(X509_Certificate cert);
+            Certificate_Entry(std::shared_ptr<Public_Key> raw_public_key);
+
+            bool has_certificate() const { return m_certificate.has_value(); }
+
+            const X509_Certificate& certificate() const;
+            std::shared_ptr<const Public_Key> public_key() const;
+
+            std::vector<uint8_t> serialize() const;
+
+            Extensions& extensions() { return m_extensions; }
+
+            const Extensions& extensions() const { return m_extensions; }
+
+         private:
+            std::optional<X509_Certificate> m_certificate;
+            std::shared_ptr<Public_Key> m_raw_public_key;
```

Comment:
Did you consider using a variant for this?

Reply:
I tried, and it certainly would work. However, the method `Certificate_Entry::public_key()` is supposed to return the credential's public key for both cases (i.e. Certificate and RawPublicKey).

Consider we have a certificate: If we'd use a variant, we would need to call `subject_public_key()` on the `X509_Certificate` every time. This returns a `std::unique_ptr<Public_Key>` and therefore re-allocates and decodes the public key for every invocation.

By splitting the members, we avoid this and decode the certificate's public key exactly once (in the constructor of `Certificate_Entry`). Albeit, it is debatable whether this is worth the additional complexity of two members.